af/fastpath/fastpath/portable_queue.py

```diff
@@ -0,0 +1,71 @@
+import six
+import platform
+import multiprocessing
+from multiprocessing.queues import Queue as BaseQueue
+
+# taken from: https://github.com/binux/pyspider/blob/d49605d6d1f07d325ff7a4ca311dc5405ed93159/pyspider/libs/multiprocessing_queue.py
+
+class SharedCounter(object):
+    """ A synchronized shared counter.
+    The locking done by multiprocessing.Value ensures that only a single
+    process or thread may read or write the in-memory ctypes object. However,
+    in order to do n += 1, Python performs a read followed by a write, so a
+    second process may read the old value before the new one is written by the
+    first process. The solution is to use a multiprocessing.Lock to guarantee
+    the atomicity of the modifications to Value.
+    This class comes almost entirely from Eli Bendersky's blog:
+    http://eli.thegreenplace.net/2012/01/04/shared-counter-with-pythons-multiprocessing/
+    """
+
+    def __init__(self, n=0):
+        self.count = multiprocessing.Value('i', n)
+
+    def increment(self, n=1):
+        """ Increment the counter by n (default = 1) """
+        with self.count.get_lock():
+            self.count.value += n
+
+    @property
+    def value(self):
+        """ Return the value of the counter """
+        return self.count.value
+
+
+class MultiProcessingQueue(BaseQueue):
```

Comment:
Wrapping a Queue instead of subclassing it sounds safer: we avoid touching the internals and doing the super() calls and we implement only the methods that are needed by the fastpath. The same wrapper can be used on every OS so that we always exercise the same code. The cost of using a lock should be negligible.

Reply:
I wasn't really sure how you were using the Queue and what plans you had for it, so it seemed safer to have an equivalent portable Queue. It's actually pretty crazy that the standard library python Queue is not portable even just between macOS and linux.